compiler/rustc_abi/src/layout.rs

```diff
@@ -934,3 +775,287 @@ pub trait LayoutCalculator {
         })
     }
 }
+
+fn univariant(
+    this: &(impl LayoutCalculator + ?Sized),
+    dl: &TargetDataLayout,
+    fields: &IndexSlice<FieldIdx, Layout<'_>>,
+    repr: &ReprOptions,
+    kind: StructKind,
+    niche_bias_start: bool,
+) -> Option<LayoutS> {
+    let pack = repr.pack;
+    let mut align = if pack.is_some() { dl.i8_align } else { dl.aggregate_align };
+    let mut inverse_memory_index: IndexVec<u32, FieldIdx> = fields.indices().collect();
+    let optimize = !repr.inhibit_struct_field_reordering_opt();
+    if optimize {
+        let end = if let StructKind::MaybeUnsized = kind { fields.len() - 1 } else { fields.len() };
+        let optimizing = &mut inverse_memory_index.raw[..end];
+
+        // If `-Z randomize-layout` was enabled for the type definition we can shuffle
+        // the field ordering to try and catch some code making assumptions about layouts
+        // we don't guarantee
+        if repr.can_randomize_type_layout() && cfg!(feature = "randomize") {
+            #[cfg(feature = "randomize")]
+            {
+                // `ReprOptions.layout_seed` is a deterministic seed that we can use to
+                // randomize field ordering with
+                let mut rng = Xoshiro128StarStar::seed_from_u64(repr.field_shuffle_seed.as_u64());
+
+                // Shuffle the ordering of the fields
+                optimizing.shuffle(&mut rng);
+            }
+            // Otherwise we just leave things alone and actually optimize the type's fields
+        } else {
+            let max_field_align = fields.iter().map(|f| f.align().abi.bytes()).max().unwrap_or(1);
+            let any_niche = fields.iter().any(|f| f.largest_niche().is_some());
+            let effective_field_align = |layout: Layout<'_>| {
+                if let Some(pack) = pack {
+                    // return the packed alignment in bytes
+                    layout.align().abi.min(pack).bytes()
+                } else {
+                    // returns log2(effective-align).
+                    // This is ok since `pack` applies to all fields equally.
+                    // The calculation assumes that size is an integer multiple of align, except for ZSTs.
+                    //
+                    // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
+                    let align = layout.align().abi.bytes();
+                    let size = layout.size().bytes();
+                    let size_as_align = align.max(size).trailing_zeros();
+                    let size_as_align = if any_niche {
+                        max_field_align.trailing_zeros().min(size_as_align)
+                    } else {
+                        size_as_align
+                    };
```

Comment:
Why is this needed? Tbh I'm confused by this whole closure...

Also it looks like you don't need to compute `max_field_align` if `!any_niche`, might be a good idea to do something like `let max_field_align = any_niche.then(|| ...);`

Reply:
Given

```
struct A(u8, [u8; 16]);

struct B(bool, [u8; 16]);
```

We want A to move the array to the front so it has a chance to be aligned when embedded in a struct with higher alignment. This allows aligned loads.

But for B we want to prioritize the bool to get a niche at offset 0.

Reply:
Start is generally preferred because it doesn't suffer from alignment padding.
End is 2nd choice and useful when it's in the largest variant in an enum.

Reply:
> Also it looks like you don't need to compute max_field_align if !any_niche, might be a good idea to do something like let max_field_align = any_niche.then(|| ...);

We make many passes over the array anyway, including sorting, so probably not worth optimizing it. And I think it would hurt readability or require additional explanation if it only gets calculated conditionally.